// @heygen/liveavatar-web-sdk/lib/index.esm.js
import { ConnectionQuality as ConnectionQuality$1, ParticipantEvent, RoomEvent, ConnectionState, Track, createLocalAudioTrack, TrackEvent, Room, VideoPresets, supportsDynacast, supportsAdaptiveStream } from 'livekit-client';
import { EventEmitter } from 'events';
import WebRTCIssueDetector from 'webrtc-issue-detector';

/******************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
/* global Reflect, Promise, SuppressedError, Symbol, Iterator */


function __awaiter(thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
}

typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
};

var SessionEvent;
(function (SessionEvent) {
    SessionEvent["SESSION_STATE_CHANGED"] = "session.state_changed";
    SessionEvent["SESSION_STREAM_READY"] = "session.stream_ready";
    SessionEvent["SESSION_CONNECTION_QUALITY_CHANGED"] = "session.connection_quality_changed";
    SessionEvent["SESSION_DISCONNECTED"] = "session.disconnected";
})(SessionEvent || (SessionEvent = {}));
var AgentEventsEnum;
(function (AgentEventsEnum) {
    AgentEventsEnum["SESSION_UPDATED"] = "session.updated";
    AgentEventsEnum["SESSION_STATE_UPDATED"] = "session.state_updated";
    AgentEventsEnum["USER_SPEAK_STARTED"] = "user.speak_started";
    AgentEventsEnum["USER_SPEAK_ENDED"] = "user.speak_ended";
    AgentEventsEnum["USER_TRANSCRIPTION"] = "user.transcription";
    AgentEventsEnum["USER_TRANSCRIPTION_CHUNK"] = "user.transcription.chunk";
    AgentEventsEnum["AVATAR_TRANSCRIPTION"] = "avatar.transcription";
    AgentEventsEnum["AVATAR_TRANSCRIPTION_CHUNK"] = "avatar.transcription.chunk";
    AgentEventsEnum["AVATAR_SPEAK_STARTED"] = "avatar.speak_started";
    AgentEventsEnum["AVATAR_SPEAK_ENDED"] = "avatar.speak_ended";
    AgentEventsEnum["ELEVENLABS_AGENT_EVENT"] = "elevenlabs_agent_event";
    AgentEventsEnum["SESSION_STOPPED"] = "session.stopped";
})(AgentEventsEnum || (AgentEventsEnum = {}));
const getAgentEventEmitArgs = (event) => {
    if ("event_type" in event) {
        switch (event.event_type) {
            case AgentEventsEnum.USER_SPEAK_STARTED: {
                const payload = {
                    event_id: event.event_id,
                    event_type: event.event_type,
                    source_event_id: event.source_event_id,
                    session_id: event.session_id,
                };
                return [AgentEventsEnum.USER_SPEAK_STARTED, payload];
            }
            case AgentEventsEnum.USER_SPEAK_ENDED: {
                const payload = {
                    event_id: event.event_id,
                    event_type: event.event_type,
                    source_event_id: event.source_event_id,
                    session_id: event.session_id,
                };
                return [AgentEventsEnum.USER_SPEAK_ENDED, payload];
            }
            case AgentEventsEnum.USER_TRANSCRIPTION: {
                const payload = {
                    event_id: event.event_id,
                    event_type: event.event_type,
                    source_event_id: event.source_event_id,
                    session_id: event.session_id,
                    text: event.text,
                };
                return [AgentEventsEnum.USER_TRANSCRIPTION, payload];
            }
            case AgentEventsEnum.USER_TRANSCRIPTION_CHUNK: {
                const payload = {
                    event_id: event.event_id,
                    event_type: event.event_type,
                    source_event_id: event.source_event_id,
                    session_id: event.session_id,
                    text: event.text,
                };
                return [AgentEventsEnum.USER_TRANSCRIPTION_CHUNK, payload];
            }
            case AgentEventsEnum.AVATAR_SPEAK_STARTED: {
                const payload = {
                    event_id: event.event_id,
                    event_type: event.event_type,
                    source_event_id: event.source_event_id,
                    session_id: event.session_id,
                };
                return [AgentEventsEnum.AVATAR_SPEAK_STARTED, payload];
            }
            case AgentEventsEnum.AVATAR_SPEAK_ENDED: {
                const payload = {
                    event_id: event.event_id,
                    event_type: event.event_type,
                    source_event_id: event.source_event_id,
                    session_id: event.session_id,
                };
                return [AgentEventsEnum.AVATAR_SPEAK_ENDED, payload];
            }
            case AgentEventsEnum.AVATAR_TRANSCRIPTION: {
                const payload = {
                    event_id: event.event_id,
                    event_type: event.event_type,
                    source_event_id: event.source_event_id,
                    session_id: event.session_id,
                    text: event.text,
                };
                return [AgentEventsEnum.AVATAR_TRANSCRIPTION, payload];
            }
            case AgentEventsEnum.AVATAR_TRANSCRIPTION_CHUNK: {
                const payload = {
                    event_id: event.event_id,
                    event_type: event.event_type,
                    source_event_id: event.source_event_id,
                    session_id: event.session_id,
                    text: event.text,
                };
                return [AgentEventsEnum.AVATAR_TRANSCRIPTION_CHUNK, payload];
            }
            case AgentEventsEnum.ELEVENLABS_AGENT_EVENT: {
                const payload = {
                    event_id: event.event_id,
                    event_type: event.event_type,
                    source_event_id: event.source_event_id,
                    session_id: event.session_id,
                    elevenlabs_event_type: event.elevenlabs_event_type,
                    data: event.data,
                };
                return [AgentEventsEnum.ELEVENLABS_AGENT_EVENT, payload];
            }
            case AgentEventsEnum.SESSION_STOPPED: {
                const payload = {
                    event_id: event.event_id,
                    event_type: event.event_type,
                    source_event_id: event.source_event_id,
                    session_id: event.session_id,
                    stop_reason: event.stop_reason,
                };
                return [AgentEventsEnum.SESSION_STOPPED, payload];
            }
            default:
                console.warn("Received event type:", event === null || event === void 0 ? void 0 : event.event_type);
                console.warn("New unsupported event type");
                return null;
        }
    }
    return null;
};
var CommandEventsEnum;
(function (CommandEventsEnum) {
    CommandEventsEnum["SESSION_UPDATE"] = "session.update";
    CommandEventsEnum["SESSION_STOP"] = "session.stop";
    CommandEventsEnum["AVATAR_INTERRUPT"] = "avatar.interrupt";
    // AVATAR_INTERRUPT_VIDEO = "avatar.interrupt_video",
    CommandEventsEnum["AVATAR_SPEAK_TEXT"] = "avatar.speak_text";
    CommandEventsEnum["AVATAR_SPEAK_RESPONSE"] = "avatar.speak_response";
    CommandEventsEnum["AVATAR_SPEAK_AUDIO"] = "avatar.speak_audio";
    CommandEventsEnum["AVATAR_START_LISTENING"] = "avatar.start_listening";
    CommandEventsEnum["AVATAR_STOP_LISTENING"] = "avatar.stop_listening";
    CommandEventsEnum["ELEVENLABS_AGENT_COMMAND"] = "elevenlabs_agent_command";
})(CommandEventsEnum || (CommandEventsEnum = {}));
var ElevenLabsAgentCommandType;
(function (ElevenLabsAgentCommandType) {
    ElevenLabsAgentCommandType["USER_MESSAGE"] = "user_message";
    ElevenLabsAgentCommandType["CONTEXTUAL_UPDATE"] = "contextual_update";
    ElevenLabsAgentCommandType["USER_ACTIVITY"] = "user_activity";
    ElevenLabsAgentCommandType["CLIENT_TOOL_RESULT"] = "client_tool_result";
})(ElevenLabsAgentCommandType || (ElevenLabsAgentCommandType = {}));

var SessionMode;
(function (SessionMode) {
    SessionMode["FULL"] = "FULL";
    SessionMode["LITE"] = "LITE";
})(SessionMode || (SessionMode = {}));
var AgentType;
(function (AgentType) {
    AgentType["FULL"] = "FULL";
    AgentType["OPENAI_REALTIME"] = "OPENAI_REALTIME";
    AgentType["ELEVENLABS_AGENT"] = "ELEVENLABS_AGENT";
    AgentType["GEMINI_REALTIME"] = "GEMINI_REALTIME";
    AgentType["UNKNOWN"] = "UNKNOWN";
})(AgentType || (AgentType = {}));
var SessionState;
(function (SessionState) {
    SessionState["INACTIVE"] = "INACTIVE";
    SessionState["CONNECTING"] = "CONNECTING";
    SessionState["CONNECTED"] = "CONNECTED";
    SessionState["DISCONNECTING"] = "DISCONNECTING";
    SessionState["DISCONNECTED"] = "DISCONNECTED";
})(SessionState || (SessionState = {}));
var SessionDisconnectReason;
(function (SessionDisconnectReason) {
    SessionDisconnectReason["UNKNOWN_REASON"] = "UNKNOWN_REASON";
    SessionDisconnectReason["CLIENT_INITIATED"] = "CLIENT_INITIATED";
    SessionDisconnectReason["SESSION_START_FAILED"] = "SESSION_START_FAILED";
    SessionDisconnectReason["SERVER_INITIATED"] = "SERVER_INITIATED";
})(SessionDisconnectReason || (SessionDisconnectReason = {}));
var Language;
(function (Language) {
    Language["af"] = "af";
    Language["sq"] = "sq";
    Language["am"] = "am";
    Language["ar"] = "ar";
    Language["hy"] = "hy";
    Language["as"] = "as";
    Language["ast"] = "ast";
    Language["az"] = "az";
    Language["ba"] = "ba";
    Language["eu"] = "eu";
    Language["be"] = "be";
    Language["bn"] = "bn";
    Language["bs"] = "bs";
    Language["br"] = "br";
    Language["bg"] = "bg";
    Language["my"] = "my";
    Language["ca"] = "ca";
    Language["ceb"] = "ceb";
    Language["zh"] = "zh";
    Language["hr"] = "hr";
    Language["cs"] = "cs";
    Language["da"] = "da";
    Language["nl"] = "nl";
    Language["en"] = "en";
    Language["et"] = "et";
    Language["fo"] = "fo";
    Language["fi"] = "fi";
    Language["fr"] = "fr";
    Language["fy"] = "fy";
    Language["ff"] = "ff";
    Language["gd"] = "gd";
    Language["gl"] = "gl";
    Language["lg"] = "lg";
    Language["ka"] = "ka";
    Language["de"] = "de";
    Language["el"] = "el";
    Language["gu"] = "gu";
    Language["ht"] = "ht";
    Language["ha"] = "ha";
    Language["haw"] = "haw";
    Language["he"] = "he";
    Language["hi"] = "hi";
    Language["hu"] = "hu";
    Language["is"] = "is";
    Language["ig"] = "ig";
    Language["ilo"] = "ilo";
    Language["id"] = "id";
    Language["ga"] = "ga";
    Language["it"] = "it";
    Language["ja"] = "ja";
    Language["jv"] = "jv";
    Language["kn"] = "kn";
    Language["kk"] = "kk";
    Language["km"] = "km";
    Language["ko"] = "ko";
    Language["lo"] = "lo";
    Language["la"] = "la";
    Language["lv"] = "lv";
    Language["lb"] = "lb";
    Language["ln"] = "ln";
    Language["lt"] = "lt";
    Language["mk"] = "mk";
    Language["mg"] = "mg";
    Language["ms"] = "ms";
    Language["ml"] = "ml";
    Language["mt"] = "mt";
    Language["mi"] = "mi";
    Language["mr"] = "mr";
    Language["mo"] = "mo";
    Language["mn"] = "mn";
    Language["ne"] = "ne";
    Language["no"] = "no";
    Language["nn"] = "nn";
    Language["oc"] = "oc";
    Language["or"] = "or";
    Language["pa"] = "pa";
    Language["ps"] = "ps";
    Language["fa"] = "fa";
    Language["pl"] = "pl";
    Language["pt"] = "pt";
    Language["ro"] = "ro";
    Language["ru"] = "ru";
    Language["sa"] = "sa";
    Language["sr"] = "sr";
    Language["sn"] = "sn";
    Language["sd"] = "sd";
    Language["si"] = "si";
    Language["sk"] = "sk";
    Language["sl"] = "sl";
    Language["so"] = "so";
    Language["es"] = "es";
    Language["su"] = "su";
    Language["sw"] = "sw";
    Language["ss"] = "ss";
    Language["sv"] = "sv";
    Language["tl"] = "tl";
    Language["tg"] = "tg";
    Language["ta"] = "ta";
    Language["tt"] = "tt";
    Language["te"] = "te";
    Language["th"] = "th";
    Language["bo"] = "bo";
    Language["tn"] = "tn";
    Language["tr"] = "tr";
    Language["tk"] = "tk";
    Language["uk"] = "uk";
    Language["ur"] = "ur";
    Language["uz"] = "uz";
    Language["vi"] = "vi";
    Language["cy"] = "cy";
    Language["wo"] = "wo";
    Language["xh"] = "xh";
    Language["yi"] = "yi";
    Language["yo"] = "yo";
    Language["zu"] = "zu";
})(Language || (Language = {}));

var ConnectionQuality;
(function (ConnectionQuality) {
    ConnectionQuality["UNKNOWN"] = "UNKNOWN";
    ConnectionQuality["GOOD"] = "GOOD";
    ConnectionQuality["BAD"] = "BAD";
})(ConnectionQuality || (ConnectionQuality = {}));

class AbstractConnectionQualityIndicator {
    constructor(onConnectionQualityChanged) {
        this._connectionQuality = ConnectionQuality.UNKNOWN;
        this.onConnectionQualityChanged = onConnectionQualityChanged;
    }
    get connectionQuality() {
        return this._connectionQuality;
    }
    handleStatsChanged() {
        const newConnectionQuality = this.calculateConnectionQuality();
        if (newConnectionQuality !== this._connectionQuality) {
            this._connectionQuality = newConnectionQuality;
            this.onConnectionQualityChanged(newConnectionQuality);
        }
    }
    start(params) {
        this.stop(true);
        this._start(params);
    }
    stop(muted = false) {
        this._stop();
        this._connectionQuality = ConnectionQuality.UNKNOWN;
        if (!muted) {
            this.onConnectionQualityChanged(ConnectionQuality.UNKNOWN);
        }
    }
}
function QualityIndicatorComposite(...configs) {
    class CombinedQualityIndicator extends AbstractConnectionQualityIndicator {
        constructor(onConnectionQualityChanged) {
            super(onConnectionQualityChanged);
            this.childTrackers = configs.map(({ getParams, TrackerClass }) => ({
                tracker: new TrackerClass(() => this.handleStatsChanged()),
                getParams,
            }));
        }
        calculateConnectionQuality() {
            const connectionQualities = this.childTrackers.map(({ tracker }) => tracker.connectionQuality);
            if (connectionQualities.some((quality) => quality === ConnectionQuality.BAD)) {
                return ConnectionQuality.BAD;
            }
            if (connectionQualities.every((quality) => quality === ConnectionQuality.UNKNOWN)) {
                return ConnectionQuality.UNKNOWN;
            }
            return ConnectionQuality.GOOD;
        }
        _start(params) {
            this.childTrackers.forEach(({ tracker, getParams }) => tracker.start(getParams(params)));
        }
        _stop() {
            this.childTrackers.forEach(({ tracker }) => tracker.stop(true));
        }
    }
    return CombinedQualityIndicator;
}

class LiveKitConnectionQualityIndicator extends AbstractConnectionQualityIndicator {
    constructor() {
        super(...arguments);
        this.room = null;
        this.liveKitConnectionQuality = ConnectionQuality$1.Unknown;
        this.liveKitConnectionState = null;
        this.handleConnectionQualityChanged = (quality) => {
            this.liveKitConnectionQuality = quality;
            this.handleStatsChanged();
        };
        this.handleConnectionStateChanged = (state) => {
            this.liveKitConnectionState = state;
            this.handleStatsChanged();
        };
    }
    _start(room) {
        this.room = room;
        this.room.localParticipant.on(ParticipantEvent.ConnectionQualityChanged, this.handleConnectionQualityChanged);
        this.room.on(RoomEvent.ConnectionStateChanged, this.handleConnectionStateChanged);
    }
    _stop() {
        if (this.room) {
            this.room.localParticipant.off(ParticipantEvent.ConnectionQualityChanged, this.handleConnectionQualityChanged);
            this.room.off(RoomEvent.ConnectionStateChanged, this.handleConnectionStateChanged);
        }
    }
    calculateConnectionQuality() {
        if ([ConnectionQuality$1.Lost, ConnectionQuality$1.Poor].includes(this.liveKitConnectionQuality)) {
            return ConnectionQuality.BAD;
        }
        if (this.liveKitConnectionState &&
            [
                ConnectionState.Disconnected,
                ConnectionState.Reconnecting,
                ConnectionState.SignalReconnecting,
            ].includes(this.liveKitConnectionState)) {
            return ConnectionQuality.BAD;
        }
        return ConnectionQuality.GOOD;
    }
}

class WebRTCConnectionQualityIndicator extends AbstractConnectionQualityIndicator {
    constructor() {
        super(...arguments);
        this.issueDetector = null;
        this.mosScores = null;
    }
    _start(peerConnection) {
        this.issueDetector = new WebRTCIssueDetector({
            autoAddPeerConnections: false,
            getStatsInterval: 3000,
            onNetworkScoresUpdated: (scores) => {
                this.mosScores = scores;
                this.handleStatsChanged();
            },
        });
        this.issueDetector.handleNewPeerConnection(peerConnection);
    }
    _stop() {
        if (this.issueDetector) {
            this.issueDetector.stopWatchingNewPeerConnections();
            this.issueDetector = null;
        }
        this.mosScores = null;
    }
    calculateConnectionQuality() {
        if (!this.mosScores ||
            (!this.mosScores.inbound && !this.mosScores.outbound)) {
            return ConnectionQuality.UNKNOWN;
        }
        if ((this.mosScores.inbound && this.mosScores.inbound < 3) ||
            (this.mosScores.outbound && this.mosScores.outbound < 3)) {
            return ConnectionQuality.BAD;
        }
        return ConnectionQuality.GOOD;
    }
}

const ConnectionQualityIndicator = QualityIndicatorComposite({
    TrackerClass: LiveKitConnectionQualityIndicator,
    getParams: (room) => room,
}, {
    TrackerClass: WebRTCConnectionQualityIndicator,
    getParams: (room) => { var _a; return ((_a = room.engine.pcManager) === null || _a === void 0 ? void 0 : _a.subscriber)._pc; },
});

var VoiceChatEvent;
(function (VoiceChatEvent) {
    VoiceChatEvent["MUTED"] = "MUTED";
    VoiceChatEvent["UNMUTED"] = "UNMUTED";
    // DEVICE_CHANGED = "DEVICE_CHANGED",
    VoiceChatEvent["STATE_CHANGED"] = "STATE_CHANGED";
})(VoiceChatEvent || (VoiceChatEvent = {}));
var PushToTalkCommandEvent;
(function (PushToTalkCommandEvent) {
    PushToTalkCommandEvent["START"] = "user.start_push_to_talk";
    PushToTalkCommandEvent["STOP"] = "user.stop_push_to_talk";
})(PushToTalkCommandEvent || (PushToTalkCommandEvent = {}));
var PushToTalkServerEvent;
(function (PushToTalkServerEvent) {
    PushToTalkServerEvent["START_SUCCESS"] = "user.push_to_talk_started";
    PushToTalkServerEvent["START_FAILED"] = "user.push_to_talk_start_failed";
    PushToTalkServerEvent["STOP_SUCCESS"] = "user.push_to_talk_stopped";
    PushToTalkServerEvent["STOP_FAILED"] = "user.push_to_talk_stop_failed";
})(PushToTalkServerEvent || (PushToTalkServerEvent = {}));

var VoiceChatState;
(function (VoiceChatState) {
    VoiceChatState["INACTIVE"] = "INACTIVE";
    VoiceChatState["STARTING"] = "STARTING";
    VoiceChatState["ACTIVE"] = "ACTIVE";
})(VoiceChatState || (VoiceChatState = {}));
var SessionInteractivityMode;
(function (SessionInteractivityMode) {
    SessionInteractivityMode["CONVERSATIONAL"] = "CONVERSATIONAL";
    SessionInteractivityMode["PUSH_TO_TALK"] = "PUSH_TO_TALK";
})(SessionInteractivityMode || (SessionInteractivityMode = {}));

const LIVEKIT_COMMAND_CHANNEL_TOPIC = "agent-control";
const LIVEKIT_SERVER_RESPONSE_CHANNEL_TOPIC = "agent-response";
const API_URL = "https://api.liveavatar.com";

const initEventPromise = (room, eventType, rejectionEventType) => {
    return new Promise((resolve, reject) => {
        const messageHandler = (roomMessage, _, __, topic) => {
            if (topic !== LIVEKIT_SERVER_RESPONSE_CHANNEL_TOPIC) {
                return;
            }
            let eventData = null;
            try {
                const messageString = new TextDecoder().decode(roomMessage);
                eventData = JSON.parse(messageString);
            }
            catch (_a) {
                return;
            }
            if (eventData && "event_type" in eventData) {
                const type = eventData.event_type;
                if (type === eventType || type === rejectionEventType) {
                    room.removeListener(RoomEvent.DataReceived, messageHandler);
                    const isRejection = type === rejectionEventType;
                    if (isRejection) {
                        reject();
                    }
                    else {
                        resolve();
                    }
                }
            }
        };
        room.on(RoomEvent.DataReceived, messageHandler);
    });
};

class VoiceChat extends EventEmitter {
    constructor(room) {
        super();
        this._state = VoiceChatState.INACTIVE;
        this.track = null;
        this.mode = null;
        this.pushToTalkStarted = false;
        this.room = room;
    }
    get isConnected() {
        return (this.room.state !== ConnectionState.Disconnected &&
            this.room.state !== ConnectionState.Connecting);
    }
    setMode(mode) {
        if (this.mode) {
            console.warn("Voice chat mode can only be set once");
            return;
        }
        this.mode = mode;
    }
    get state() {
        return this._state;
    }
    get isMuted() {
        var _a, _b;
        return (_b = (_a = this.track) === null || _a === void 0 ? void 0 : _a.isMuted) !== null && _b !== void 0 ? _b : true;
    }
    start() {
        return __awaiter(this, arguments, void 0, function* (config = {}) {
            if (!this.isConnected) {
                console.warn("Voice chat can only be started when session is active");
                return;
            }
            if (this._state !== VoiceChatState.INACTIVE) {
                console.warn("Voice chat is already started");
                return;
            }
            this.state = VoiceChatState.STARTING;
            const { defaultMuted, deviceId, mode } = config;
            if (mode) {
                this.setMode(mode);
            }
            try {
                this.track = yield createLocalAudioTrack({
                    echoCancellation: true,
                    noiseSuppression: true,
                    autoGainControl: true,
                    deviceId,
                });
            }
            catch (error) {
                this.state = VoiceChatState.INACTIVE;
                throw error;
            }
            if (defaultMuted) {
                yield this.track.mute();
                this.emit(VoiceChatEvent.MUTED);
            }
            else {
                this.emit(VoiceChatEvent.UNMUTED);
            }
            yield this.room.localParticipant.publishTrack(this.track);
            this.track.on(TrackEvent.Muted, () => {
                this.emit(VoiceChatEvent.MUTED);
            });
            this.track.on(TrackEvent.Unmuted, () => {
                this.emit(VoiceChatEvent.UNMUTED);
            });
            this.state = VoiceChatState.ACTIVE;
        });
    }
    stop() {
        this.room.localParticipant.getTrackPublications().forEach((publication) => {
            if (publication.track && publication.track.kind === Track.Kind.Audio) {
                publication.track.stop();
            }
        });
        if (this.track) {
            this.track.removeAllListeners();
            this.track.stop();
            this.track = null;
        }
        this.state = VoiceChatState.INACTIVE;
    }
    mute() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.assertActive("Voice chat can only be muted when active")) {
                return;
            }
            if (this.track) {
                yield this.track.mute();
            }
        });
    }
    unmute() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.assertActive("Voice chat can only be unmuted when active")) {
                return;
            }
            if (this.track) {
                yield this.track.unmute();
            }
        });
    }
    setDevice(deviceId) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.assertActive("Voice chat device can only be set when active")) {
                return false;
            }
            if (this.track) {
                return this.track.setDeviceId(deviceId);
            }
            return false;
        });
    }
    startPushToTalk() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.assertActive("Push to talk can only be started when voice chat is active")) {
                return;
            }
            console.error("Session interactivity mode", this.mode);
            if (this.mode !== SessionInteractivityMode.PUSH_TO_TALK) {
                console.warn("Push to talk can only be started in push to talk mode");
                return;
            }
            if (this.pushToTalkStarted) {
                console.warn("Push to talk has already been started");
                return;
            }
            this.pushToTalkStarted = true;
            const promise = initEventPromise(this.room, PushToTalkServerEvent.START_SUCCESS, PushToTalkServerEvent.START_FAILED);
            this.sendPushToTalkCommand(PushToTalkCommandEvent.START);
            try {
                yield promise;
                yield this.unmute();
            }
            catch (e) {
                console.error("Failed to start push to talk", e);
                this.pushToTalkStarted = false;
                throw e;
            }
        });
    }
    stopPushToTalk() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.pushToTalkStarted) {
                console.warn("Push to talk has not been started");
                return;
            }
            const promise = initEventPromise(this.room, PushToTalkServerEvent.STOP_SUCCESS, PushToTalkServerEvent.STOP_FAILED);
            this.sendPushToTalkCommand(PushToTalkCommandEvent.STOP);
            try {
                yield promise;
                this.pushToTalkStarted = false;
            }
            catch (e) {
                console.error("Failed to stop push to talk", e);
                throw e;
            }
        });
    }
    sendPushToTalkCommand(command) {
        const data = new TextEncoder().encode(JSON.stringify({ event_type: command }));
        this.room.localParticipant.publishData(data, {
            reliable: true,
            topic: LIVEKIT_COMMAND_CHANNEL_TOPIC,
        });
    }
    set state(state) {
        if (this._state !== state) {
            this._state = state;
            this.emit(VoiceChatEvent.STATE_CHANGED, state);
        }
    }
    assertActive(warnMessage) {
        if (this.state !== VoiceChatState.ACTIVE) {
            console.warn(warnMessage !== null && warnMessage !== void 0 ? warnMessage : "Voice chat is not active");
            return false;
        }
        return true;
    }
}

const DEFAULT_ERROR_CODE = 500;
const SUCCESS_CODE = 1000;
class SessionApiError extends Error {
    constructor(message, errorCode, status) {
        super(message);
        this.status = null;
        this.errorCode = errorCode !== null && errorCode !== void 0 ? errorCode : DEFAULT_ERROR_CODE;
        this.status = status !== null && status !== void 0 ? status : null;
    }
}
class SessionAPIClient {
    constructor(sessionToken, apiUrl = API_URL) {
        this.sessionToken = sessionToken;
        this.apiUrl = apiUrl !== null && apiUrl !== void 0 ? apiUrl : API_URL;
    }
    request(path, params) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const response = yield fetch(`${this.apiUrl}${path}`, Object.assign(Object.assign({}, params), { credentials: "include", headers: Object.assign({ Authorization: `Bearer ${this.sessionToken}`, "Content-Type": "application/json" }, params.headers) }));
                if (!response.ok) {
                    const data = yield response.json();
                    throw new SessionApiError(data.message || `API request failed with status ${response.status}`, data.code, response.status);
                }
                const data = yield response.json();
                if (data.code !== SUCCESS_CODE) {
                    throw new SessionApiError(data.message || "API request failed");
                }
                return data.data;
            }
            catch (err) {
                if (err instanceof SessionApiError) {
                    throw err;
                }
                throw new SessionApiError("API request failed");
            }
        });
    }
    startSession() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.request(`/v1/sessions/start`, { method: "POST" });
        });
    }
    stopSession() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.request(`/v1/sessions/stop`, { method: "POST" });
        });
    }
    keepAlive() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield this.request(`/v1/sessions/keep-alive`, { method: "POST" });
        });
    }
}

/**
 * Splits a PCM 24KHz audio string (raw 16-bit signed PCM) into chunks.
 *
 * The first chunk is 400ms to minimize time-to-first-audio; subsequent chunks
 * are 1s to reduce message overhead once playback is underway.
 *
 * At 24,000Hz, 16-bit mono: 1 sample = 2 bytes.
 *   400ms => 24000 * 0.4 * 2 = 19200 bytes
 *   1000ms => 24000 * 1.0 * 2 = 48000 bytes
 */
function splitPcm24kStringToChunks(pcmString) {
    const firstChunkBytes = 24000 * 0.4 * 2; // 19200 bytes == 400ms
    const subsequentChunkBytes = 24000 * 1.0 * 2; // 48000 bytes == 1s
    const totalLength = pcmString.length;
    const result = [];
    if (totalLength === 0)
        return result;
    result.push(pcmString.slice(0, firstChunkBytes));
    for (let i = firstChunkBytes; i < totalLength; i += subsequentChunkBytes) {
        result.push(pcmString.slice(i, i + subsequentChunkBytes));
    }
    return result;
}

const HEYGEN_PARTICIPANT_ID = "heygen";
const LIVEAVATAR_AGENT_PARTICIPANT_ID_PREFIX = "liveavatar-agent-";
const REQUIRED_PARTICIPANTS_TIMEOUT_MS = 30000;
function decodeSessionTokenPayload(token) {
    try {
        const parts = token.split(".");
        if (parts.length < 2) {
            console.warn("[decodeSessionTokenPayload] Token does not look like a JWT (parts.length =", parts.length, "). Raw token (first 40 chars):", token.slice(0, 40));
            return null;
        }
        const payload = parts[1];
        const decoded = JSON.parse(atob(payload.replace(/-/g, "+").replace(/_/g, "/")));
        return decoded;
    }
    catch (e) {
        console.warn("[decodeSessionTokenPayload] Failed to decode session token payload. Raw token (first 40 chars):", token.slice(0, 40), "Error:", e);
        return null;
    }
}
function parseSessionModeFromToken(token) {
    var _a;
    const decoded = decodeSessionTokenPayload(token);
    const mode = (_a = decoded === null || decoded === void 0 ? void 0 : decoded.start_session_data) === null || _a === void 0 ? void 0 : _a.mode;
    if (mode === SessionMode.LITE)
        return SessionMode.LITE;
    if (mode === SessionMode.FULL)
        return SessionMode.FULL;
    return SessionMode.FULL;
}
// TODO(LA-1410): replace heuristic once backend exposes explicit agent_type
// claim. Today we infer ELEVENLABS_AGENT from presence of
// elevenlabs_agent_config (and similar for other realtime providers).
function parseAgentTypeFromToken(token) {
    const decoded = decodeSessionTokenPayload(token);
    if (!decoded)
        return AgentType.UNKNOWN;
    const data = decoded.start_session_data;
    if (!data)
        return AgentType.UNKNOWN;
    // Prefer explicit agent_type claim if backend exposes one.
    if (typeof data.agent_type === "string") {
        const explicit = data.agent_type.toUpperCase();
        if (explicit in AgentType)
            return explicit;
    }
    if (data.elevenlabs_agent_config)
        return AgentType.ELEVENLABS_AGENT;
    if (data.openai_realtime_config)
        return AgentType.OPENAI_REALTIME;
    if (data.gemini_realtime_config)
        return AgentType.GEMINI_REALTIME;
    return AgentType.FULL;
}
class LiveAvatarSession extends EventEmitter {
    constructor(sessionAccessToken, config) {
        super();
        this.connectionQualityIndicator = new ConnectionQualityIndicator((quality) => this.emit(SessionEvent.SESSION_CONNECTION_QUALITY_CHANGED, quality));
        this._sessionInfo = null;
        this._sessionEventSocket = null;
        this._state = SessionState.INACTIVE;
        this._remoteAudioTrack = null;
        this._remoteVideoTrack = null;
        this._mode = parseSessionModeFromToken(sessionAccessToken);
        this._agentType = parseAgentTypeFromToken(sessionAccessToken);
        // Required to construct the room
        this.config = config !== null && config !== void 0 ? config : {};
        this.sessionClient = new SessionAPIClient(sessionAccessToken, this.config.apiUrl);
        this.room = new Room({
            adaptiveStream: supportsAdaptiveStream()
                ? {
                    pauseVideoInBackground: false,
                }
                : false,
            dynacast: supportsDynacast(),
            videoCaptureDefaults: {
                resolution: VideoPresets.h720.resolution,
            },
        });
        this._voiceChat = new VoiceChat(this.room);
        if (this.config.voiceChat &&
            typeof this.config.voiceChat === "object" &&
            this.config.voiceChat.mode) {
            this._voiceChat.setMode(this.config.voiceChat.mode);
        }
    }
    get state() {
        return this._state;
    }
    get mode() {
        return this._mode;
    }
    get agentType() {
        return this._agentType;
    }
    get sessionId() {
        var _a, _b;
        return (_b = (_a = this._sessionInfo) === null || _a === void 0 ? void 0 : _a.session_id) !== null && _b !== void 0 ? _b : null;
    }
    get connectionQuality() {
        return this.connectionQualityIndicator.connectionQuality;
    }
    get voiceChat() {
        return this._voiceChat;
    }
    get maxSessionDuration() {
        var _a, _b;
        return (_b = (_a = this._sessionInfo) === null || _a === void 0 ? void 0 : _a.max_session_duration) !== null && _b !== void 0 ? _b : null;
    }
    start() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.state !== SessionState.INACTIVE) {
                console.warn("Session is already started");
                return;
            }
            try {
                this.state = SessionState.CONNECTING;
                this._sessionInfo = yield this.sessionClient.startSession();
                const livekitRoomUrl = this._sessionInfo.livekit_url;
                const livekitClientToken = this._sessionInfo.livekit_client_token;
                const websocketUrl = this._sessionInfo.ws_url;
                // Connect to LiveKit room if provided
                if (livekitRoomUrl && livekitClientToken) {
                    // Track the different events from the room, server, and websocket
                    this.trackEvents();
                    yield this.room.connect(livekitRoomUrl, livekitClientToken);
                    yield this.waitForRequiredParticipants();
                    this.connectionQualityIndicator.start(this.room);
                }
                // Connect to WebSocket if provided
                if (websocketUrl) {
                    yield this.connectWebSocket(websocketUrl);
                    this.setupWebSocketManagement();
                }
                // Run configurations as needed
                yield this.configureSession();
                this.state = SessionState.CONNECTED;
            }
            catch (error) {
                console.error("Session start failed:", error);
                this.cleanup();
                this.postStop(SessionDisconnectReason.SESSION_START_FAILED);
                throw error;
            }
        });
    }
    stop() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.assertConnected()) {
                return;
            }
            this.state = SessionState.DISCONNECTING;
            this.cleanup();
            this.postStop(SessionDisconnectReason.CLIENT_INITIATED);
        });
    }
    keepAlive() {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.assertConnected()) {
                return;
            }
            try {
                this.sessionClient.keepAlive();
            }
            catch (error) {
                console.error("Session keep alive error on server:", error);
                throw error;
            }
        });
    }
    attach(element) {
        if (!this._remoteVideoTrack || !this._remoteAudioTrack) {
            console.warn("Stream is not yet ready");
            return;
        }
        this._remoteVideoTrack.attach(element);
        this._remoteAudioTrack.attach(element);
    }
    message(message) {
        if (!this.assertConnected()) {
            throw new Error("Session needs to be connected to send command event");
        }
        const event_id = this.generateEventId();
        const data = {
            event_id: event_id,
            event_type: CommandEventsEnum.AVATAR_SPEAK_RESPONSE,
            text: message,
        };
        this.sendCommandEvent(data);
        return event_id;
    }
    repeat(message) {
        if (!this.assertConnected()) {
            throw new Error("Session needs to be connected to send command event");
        }
        const event_id = this.generateEventId();
        const data = {
            event_id: event_id,
            event_type: CommandEventsEnum.AVATAR_SPEAK_TEXT,
            text: message,
        };
        console.warn("sending repeat command event", data);
        this.sendCommandEvent(data);
        return event_id;
    }
    repeatAudio(audio) {
        if (!this.assertConnected()) {
            throw new Error("Session needs to be connected to send command event");
        }
        if (!this._sessionEventSocket) {
            console.warn("Cannot repeat audio. Please check you're using a supported mode.");
            throw new Error("Session needs to be connected to send command event");
        }
        const event_id = this.generateEventId();
        const data = {
            event_id: event_id,
            event_type: CommandEventsEnum.AVATAR_SPEAK_AUDIO,
            audio: audio,
        };
        this.sendCommandEvent(data);
        return event_id;
    }
    startListening() {
        if (!this.assertConnected()) {
            throw new Error("Session needs to be connected to send command event");
        }
        const event_id = this.generateEventId();
        const data = {
            event_id: event_id,
            event_type: CommandEventsEnum.AVATAR_START_LISTENING,
        };
        this.sendCommandEvent(data);
        return event_id;
    }
    stopListening() {
        if (!this.assertConnected()) {
            throw new Error("Session needs to be connected to send command event");
        }
        const event_id = this.generateEventId();
        const data = {
            event_type: CommandEventsEnum.AVATAR_STOP_LISTENING,
            event_id: event_id,
        };
        this.sendCommandEvent(data);
        return event_id;
    }
    interrupt() {
        if (!this.assertConnected()) {
            throw new Error("Session needs to be connected to send command event");
        }
        const data = {
            event_type: CommandEventsEnum.AVATAR_INTERRUPT,
        };
        this.sendCommandEvent(data);
    }
    trackEvents() {
        const mediaStream = new MediaStream();
        this.room.on(RoomEvent.TrackSubscribed, (track, _publication, participant) => {
            // We need to actively track the HeyGen participant's tracks
            if (participant.identity !== HEYGEN_PARTICIPANT_ID) {
                return;
            }
            if (track.kind === "video" || track.kind === "audio") {
                if (track.kind === "video") {
                    this._remoteVideoTrack = track;
                }
                else {
                    this._remoteAudioTrack = track;
                }
                mediaStream.addTrack(track.mediaStreamTrack);
                const hasVideoTrack = mediaStream.getVideoTracks().length > 0;
                const hasAudioTrack = mediaStream.getAudioTracks().length > 0;
                if (hasVideoTrack && hasAudioTrack) {
                    this.emit(SessionEvent.SESSION_STREAM_READY);
                }
            }
        });
        this.room.on(RoomEvent.DataReceived, (roomMessage, _, __, topic) => {
            if (topic !== LIVEKIT_SERVER_RESPONSE_CHANNEL_TOPIC) {
                return;
            }
            let eventMsg = null;
            try {
                const messageString = new TextDecoder().decode(roomMessage);
                eventMsg = JSON.parse(messageString);
            }
            catch (e) {
                console.error(e);
            }
            if (!eventMsg) {
                return;
            }
            const emitArgs = getAgentEventEmitArgs(eventMsg);
            if (emitArgs) {
                const [event_type, ...event_data] = emitArgs;
                this.emit(event_type, ...event_data);
            }
        });
        this.on(AgentEventsEnum.SESSION_STOPPED, (event) => {
            console.warn("[SDK:SESSION_STOPPED] Server stopped session, reason:", event.stop_reason);
            this.cleanup();
            this.postStop(SessionDisconnectReason.SERVER_INITIATED);
        });
        this.room.on(RoomEvent.ParticipantConnected, (participant) => {
            console.warn("participantConnected", participant.identity);
        });
        this.room.on(RoomEvent.TrackUnsubscribed, (track) => {
            console.warn("trackUnsubscribed", track);
            const mediaTrack = track.mediaStreamTrack;
            if (mediaTrack) {
                mediaStream.removeTrack(mediaTrack);
            }
        });
        this.room.on(RoomEvent.Disconnected, (reason) => {
            console.warn("Room disconnected, reason:", reason);
            this.handleRoomDisconnect();
        });
        this.room.on(RoomEvent.TrackPublished, (track) => {
            console.warn("trackPublished", track);
        });
    }
    waitForRequiredParticipants() {
        var _a;
        if (this._mode !== SessionMode.FULL) {
            return Promise.resolve();
        }
        const sessionId = (_a = this._sessionInfo) === null || _a === void 0 ? void 0 : _a.session_id;
        if (!sessionId) {
            return Promise.resolve();
        }
        const agentId = `${LIVEAVATAR_AGENT_PARTICIPANT_ID_PREFIX}${sessionId}`;
        const required = new Set([HEYGEN_PARTICIPANT_ID, agentId]);
        for (const p of this.room.remoteParticipants.values()) {
            required.delete(p.identity);
        }
        if (required.size === 0) {
            return Promise.resolve();
        }
        return new Promise((resolve, reject) => {
            const onConnected = (participant) => {
                required.delete(participant.identity);
                if (required.size === 0) {
                    cleanup();
                    resolve();
                }
            };
            const timeout = setTimeout(() => {
                cleanup();
                reject(new Error(`Timed out waiting for required participants: ${Array.from(required).join(", ")}`));
            }, REQUIRED_PARTICIPANTS_TIMEOUT_MS);
            const cleanup = () => {
                clearTimeout(timeout);
                this.room.off(RoomEvent.ParticipantConnected, onConnected);
            };
            this.room.on(RoomEvent.ParticipantConnected, onConnected);
        });
    }
    connectWebSocket(websocketUrl) {
        return __awaiter(this, void 0, void 0, function* () {
            return new Promise((resolve, _reject) => {
                this._sessionEventSocket = new WebSocket(websocketUrl);
                this._sessionEventSocket.onopen = () => {
                    resolve();
                };
            });
        });
    }
    setupWebSocketManagement() {
        if (!this._sessionEventSocket) {
            return;
        }
        this._sessionEventSocket.onmessage = (event) => {
            this.handleWebSocketMessage(event);
        };
        this._sessionEventSocket.onerror = (error) => {
            console.error("WebSocket error:", error);
        };
        this._sessionEventSocket.onclose = (event) => {
            console.warn("WebSocket closed - code:", event.code, "reason:", event.reason, "wasClean:", event.wasClean);
            this.handleWebSocketDisconnect();
        };
    }
    handleWebSocketMessage(event) {
        let eventData = null;
        try {
            eventData = JSON.parse(event.data);
        }
        catch (e) {
            console.error("Failed to parse WebSocket message:", e);
            return;
        }
        if (!eventData) {
            return;
        }
        const { type, event_id } = eventData;
        if (type === "agent.speak_started") {
            this.emit(AgentEventsEnum.AVATAR_SPEAK_STARTED, {
                event_type: AgentEventsEnum.AVATAR_SPEAK_STARTED,
                event_id: event_id,
            });
        }
        else if (type === "agent.speak_ended") {
            this.emit(AgentEventsEnum.AVATAR_SPEAK_ENDED, {
                event_type: AgentEventsEnum.AVATAR_SPEAK_ENDED,
                event_id: event_id,
            });
        }
    }
    handleWebSocketDisconnect() {
        if (this.state === SessionState.DISCONNECTING ||
            this.state === SessionState.DISCONNECTED) {
            return;
        }
        if (this._sessionEventSocket &&
            this._sessionEventSocket.readyState === WebSocket.OPEN) {
            this._sessionEventSocket.close();
        }
        this._sessionEventSocket = null;
        this.cleanup();
        this.postStop(SessionDisconnectReason.UNKNOWN_REASON);
    }
    configureSession() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.config.voiceChat) {
                try {
                    yield this.voiceChat.start(typeof this.config.voiceChat === "boolean"
                        ? {}
                        : this.config.voiceChat);
                }
                catch (error) {
                    console.warn("Failed to start voice chat (microphone may be unavailable):", error);
                }
            }
        });
    }
    set state(state) {
        if (this._state === state) {
            return;
        }
        this._state = state;
        this.emit(SessionEvent.SESSION_STATE_CHANGED, state);
    }
    cleanup() {
        return __awaiter(this, void 0, void 0, function* () {
            this.connectionQualityIndicator.stop();
            this.voiceChat.stop();
            if (this._remoteAudioTrack) {
                this._remoteAudioTrack.stop();
            }
            if (this._remoteVideoTrack) {
                this._remoteVideoTrack.stop();
            }
            this._remoteAudioTrack = null;
            this._remoteVideoTrack = null;
            this.room.localParticipant.removeAllListeners();
            this.room.removeAllListeners();
            // Clean up WebSocket
            if (this._sessionEventSocket) {
                // Remove event listeners to prevent callbacks during cleanup
                this._sessionEventSocket.onopen = null;
                this._sessionEventSocket.onmessage = null;
                this._sessionEventSocket.onerror = null;
                this._sessionEventSocket.onclose = null;
                if (this._sessionEventSocket.readyState === WebSocket.OPEN ||
                    this._sessionEventSocket.readyState === WebSocket.CONNECTING) {
                    this._sessionEventSocket.close();
                }
                this._sessionEventSocket = null;
            }
            // Disconnect from room if connected
            if (this.room.state === "connected") {
                this.room.disconnect();
            }
            // Kill the session on the server
            yield this.sessionClient.stopSession();
        });
    }
    postStop(reason) {
        this.state = SessionState.DISCONNECTED;
        this.emit(SessionEvent.SESSION_DISCONNECTED, reason);
    }
    handleRoomDisconnect() {
        this.cleanup();
        this.postStop(SessionDisconnectReason.UNKNOWN_REASON);
    }
    sendCommandEvent(commandEvent) {
        // Use WebSocket if available, otherwise use LiveKit data channel
        if (this._sessionEventSocket &&
            this._sessionEventSocket.readyState === WebSocket.OPEN) {
            this.sendCommandEventToWebSocket(commandEvent);
        }
        else if (this.room.state === "connected") {
            this.publishAgentControl(commandEvent);
        }
        else {
            console.warn("No active connection to send command event");
        }
    }
    publishAgentControl(payload) {
        if (this.room.state !== "connected") {
            throw new Error("LiveKit room not connected — cannot publish agent-control payload");
        }
        const data = new TextEncoder().encode(JSON.stringify(payload));
        this.room.localParticipant.publishData(data, {
            reliable: true,
            topic: LIVEKIT_COMMAND_CHANNEL_TOPIC,
        });
    }
    generateEventId() {
        // Use native browser crypto API
        if (typeof crypto !== "undefined" && crypto.randomUUID) {
            return crypto.randomUUID();
        }
        // Fallback for older browsers
        return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
            const r = (Math.random() * 16) | 0;
            const v = c === "x" ? r : (r & 0x3) | 0x8;
            return v.toString(16);
        });
    }
    sendCommandEventToWebSocket(commandEvent) {
        if (!this._sessionEventSocket ||
            this._sessionEventSocket.readyState !== WebSocket.OPEN) {
            console.warn("WebSocket not open to send command event");
            return;
        }
        const event_type = commandEvent.event_type;
        const event_id = this.generateEventId();
        let audioChunks = [];
        switch (event_type) {
            case CommandEventsEnum.AVATAR_SPEAK_AUDIO:
                audioChunks = splitPcm24kStringToChunks(commandEvent.audio);
                for (const audioChunk of audioChunks) {
                    this._sessionEventSocket.send(JSON.stringify({
                        type: "agent.speak",
                        event_id: event_id,
                        audio: audioChunk,
                    }));
                }
                this._sessionEventSocket.send(JSON.stringify({
                    type: "agent.speak_end",
                    event_id: event_id,
                }));
                return;
            case CommandEventsEnum.AVATAR_SPEAK_TEXT:
            case CommandEventsEnum.AVATAR_SPEAK_RESPONSE:
                throw new Error("Not permitted in LITE mode");
            case CommandEventsEnum.AVATAR_INTERRUPT:
                this._sessionEventSocket.send(JSON.stringify({
                    type: "agent.interrupt",
                    event_id: event_id,
                }));
                return;
            case CommandEventsEnum.AVATAR_START_LISTENING:
                this._sessionEventSocket.send(JSON.stringify({
                    type: "agent.start_listening",
                    event_id: event_id,
                }));
                return;
            case CommandEventsEnum.AVATAR_STOP_LISTENING:
                this._sessionEventSocket.send(JSON.stringify({
                    type: "agent.stop_listening",
                    event_id: event_id,
                }));
                return;
            default:
                console.warn("Unsupported command event type:", event_type);
                break;
        }
    }
    assertConnected() {
        if (this.state !== SessionState.CONNECTED) {
            console.warn("Session is not connected");
            return false;
        }
        return true;
    }
}

class ElevenLabsAgentSessionError extends Error {
    constructor(message) {
        super(message);
        this.name = "ElevenLabsAgentSessionError";
    }
}
/**
 * Session class for LiveAvatar sessions backed by an ElevenLabs Agent.
 *
 * Inherits the connection lifecycle, voice chat, and inbound event surface
 * from LiveAvatarSession. Adds outbound commands that wrap ElevenLabs
 * client-event types (user_message, contextual_update, user_activity,
 * client_tool_result) and ship them on the LiveKit `agent-control` topic.
 *
 * Refuses to send if the JWT does not indicate an ELEVENLABS agent.
 *
 * NOTE: The non-ElevenLabs LiveAvatar command methods (`message`, `repeat`,
 * `repeatAudio`) are not supported on this path and will throw. In a future
 * iteration we may map these onto ElevenLabs equivalents (e.g. `message` ->
 * `sendUserMessage`) so a developer can swap LiveAvatarSession for
 * ElevenLabsAgentSession with no call-site changes.
 */
class ElevenLabsAgentSession extends LiveAvatarSession {
    constructor(sessionAccessToken, config) {
        super(sessionAccessToken, config);
        // TODO(LA-1410): Re-enable agent_type validation once the backend echoes
        // `start_session_data.agent_type = "ELEVENLABS_AGENT"` in the JWT. Today
        // the token omits agent_type entirely, so any client-side check is
        // unreliable. We trust the caller's choice of class — the server still
        // enforces the real agent type and silently drops mismatched commands.
        //
        // if (this.agentType !== AgentType.ELEVENLABS_AGENT) {
        //   console.warn(
        //     `[ElevenLabsAgentSession] Token agent_type detected as "${this.agentType}", ` +
        //       `expected "${AgentType.ELEVENLABS_AGENT}".`,
        //   );
        // }
    }
    /**
     * Low-level: send any whitelisted ElevenLabs agent command. Prefer the
     * named convenience methods below.
     */
    sendElevenLabsAgentCommand(payload, options) {
        var _a;
        if (!this.assertConnected()) {
            throw new ElevenLabsAgentSessionError("Session must be connected before sending ElevenLabs agent commands");
        }
        const sessionId = this.sessionId;
        if (!sessionId) {
            throw new ElevenLabsAgentSessionError("Session id is not yet available — wait for session.started");
        }
        const event_id = this.generateEventId();
        const wrapper = Object.assign({ event_type: CommandEventsEnum.ELEVENLABS_AGENT_COMMAND, event_id, session_id: sessionId, source_event_id: (_a = options === null || options === void 0 ? void 0 : options.sourceEventId) !== null && _a !== void 0 ? _a : null }, payload);
        this.publishAgentControl(wrapper);
        return event_id;
    }
    /** Send a user message — equivalent to typing in the ElevenLabs widget. */
    sendUserMessage(text) {
        return this.sendElevenLabsAgentCommand({
            elevenlabs_event_type: ElevenLabsAgentCommandType.USER_MESSAGE,
            data: { text },
        });
    }
    /** Send a contextual update — silent context the agent can use. */
    sendContextualUpdate(text) {
        return this.sendElevenLabsAgentCommand({
            elevenlabs_event_type: ElevenLabsAgentCommandType.CONTEXTUAL_UPDATE,
            data: { text },
        });
    }
    /** Keepalive / typing-indicator ping. */
    sendUserActivity() {
        return this.sendElevenLabsAgentCommand({
            elevenlabs_event_type: ElevenLabsAgentCommandType.USER_ACTIVITY,
        });
    }
    /** Reply to a client_tool_call inbound event with the tool result. */
    sendClientToolResult(args) {
        var _a;
        return this.sendElevenLabsAgentCommand({
            elevenlabs_event_type: ElevenLabsAgentCommandType.CLIENT_TOOL_RESULT,
            data: {
                tool_call_id: args.toolCallId,
                result: args.result,
                is_error: (_a = args.isError) !== null && _a !== void 0 ? _a : false,
            },
        }, { sourceEventId: args.sourceEventId });
    }
    // ---------------------------------------------------------------------
    // Disabled LiveAvatar command surface
    //
    // TODO(LA-1410-followup): Remap these onto the ElevenLabs equivalents so
    // a consumer can swap LiveAvatarSession <-> ElevenLabsAgentSession with
    // no call-site changes. Likely mapping:
    //   message(text)     -> sendUserMessage(text)
    //   repeat(text)      -> sendContextualUpdate(text)  (or a new EL event)
    //   repeatAudio(b64)  -> not supported (no analogue)
    // ---------------------------------------------------------------------
    message(_message) {
        throw new ElevenLabsAgentSessionError("message() is not supported on ElevenLabsAgentSession — use sendUserMessage()");
    }
    repeat(_message) {
        throw new ElevenLabsAgentSessionError("repeat() is not supported on ElevenLabsAgentSession — the ElevenLabs agent owns response generation");
    }
    repeatAudio(_audio) {
        throw new ElevenLabsAgentSessionError("repeatAudio() is not supported on ElevenLabsAgentSession");
    }
}

export { AgentEventsEnum, AgentType, CommandEventsEnum, ConnectionQuality, ElevenLabsAgentCommandType, ElevenLabsAgentSession, ElevenLabsAgentSessionError, Language, LiveAvatarSession, SessionDisconnectReason, SessionEvent, SessionInteractivityMode, SessionMode, SessionState, VoiceChatEvent, VoiceChatState, parseAgentTypeFromToken };
